### PR TITLE
Use Kryo serialization for Spark processing

### PR DESCRIPTION
**Pull request by tomekl007**
It switches our spark serialization to kryo which provides better performance and lower memory footprint (https://spark.apache.org/docs/latest/tuning.html#:~:text=Kryo%20is%20significantly%20faster%20and,in%20advance%20for%20best%20performance) compared to standard java serialization.
The `ReferencesKryoRegistrator` is responsible for registering all the classes:
https://github.com/projectnessie/nessie/pull/3599/files#diff-eb8af2f0c5b82bcb9caa73ac36f984925c668b78b804ddf3324a85e07875db14R42

The Spark Session that uses this registration should be started with `spark.kryo.registrationRequired = true` to reduce the memory footprint of the processing. For more explanation, see the docs:
https://github.com/projectnessie/nessie/pull/3599/files#diff-eb8af2f0c5b82bcb9caa73ac36f984925c668b78b804ddf3324a85e07875db14R45-R55

To see how the `sparkSession` should be created, see: https://github.com/projectnessie/nessie/pull/3599/files#diff-10adbf0faca7bef7d4f4326aece9753dd532b5e64f8864f924ad981a2cff46a7R60-R74

This PR also refactors how the `SparkSession` is created and closed in a test - it reuses the same session and closes after all tests. It also simplifies the spark processing to make it stateless, for more info see: https://github.com/projectnessie/nessie/pull/3599/files#r825759507

### CODE REVIEW

**Review by snazy — COMMENTED** [review]
Didn't look deeply into this change, but I think only `Reference` and its subtypes need to be serialized - probably also even just the reference type, name and hash.
So maybe just use a custom serializer for the necessary fields.

**Comment by tomekl007** [comment]
> Didn't look deeply into this change, but I think only `Reference` and its subtypes need to be serialized - probably also even just the reference type, name and hash. So maybe just use a custom serializer for the necessary fields.

Yes, doing it here:
https://github.com/projectnessie/nessie/pull/3599/files#diff-eb8af2f0c5b82bcb9caa73ac36f984925c668b78b804ddf3324a85e07875db14R48-R52

I am hitting this now:
```

org.apache.spark.SparkException: Task not serializable

	at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:416)
	at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:406)
	at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:162)
	at org.apache.spark.SparkContext.clean(SparkContext.scala:2459)
	at org.apache.spark.rdd.RDD.$anonfun$map$1(RDD.scala:422)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:414)
	at org.apache.spark.rdd.RDD.map(RDD.scala:421)
	at org.apache.spark.api.java.JavaRDDLike.map(JavaRDDLike.scala:93)
	at org.apache.spark.api.java.JavaRDDLike.map$(JavaRDDLike.scala:92)
	at org.apache.spark.api.java.AbstractJavaRDDLike.map(JavaRDDLike.scala:45)
	at org.projectnessie.gc.base.DistributedIdentifyContents.getLiveContentsBloomFilters(DistributedIdentifyContents.java:61)
	at org.projectnessie.gc.base.GCImpl.identifyExpiredContents(GCImpl.java:98)
	at org.projectnessie.jaxrs.gc.AbstractRestGC.performGc(AbstractRestGC.java:139)
	at org.projectnessie.jaxrs.gc.AbstractRestGCTest.testSingleRefDeadRefCutoff(AbstractRestGCTest.java:478)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.io.NotSerializableException: org.projectnessie.model.ImmutableBranch
Serialization stack:
	- object not serializable (class: org.projectnessie.model.ImmutableBranch, value: Branch{name=singleRefDeadRefCutoff, metadata=null, hash=38e5e7901d4f20560b86d2cdc761b279e47e7d7555a145f6edc43477a424ed23})
	- writeObject data (class: java.util.HashMap)
	- object (class java.util.HashMap, {Branch{name=singleRefDeadRefCutoff, metadata=null, hash=38e5e7901d4f20560b86d2cdc761b279e47e7d7555a145f6edc43477a424ed23}=2022-03-11T14:04:02.821939Z})
	- element of array (index: 1)
	- array (class [Ljava.lang.Object;, size 3)
	- field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
	- object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.projectnessie.gc.base.DistributedIdentifyContents, functionalInterfaceMethod=org/apache/spark/api/java/function/Function.call:(Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/projectnessie/gc/base/DistributedIdentifyContents.lambda$getLiveContentsBloomFilters$98ea120e$1:(JLjava/util/Map;Lorg/projectnessie/gc/base/GCParams;Lorg/projectnessie/model/Reference;)Ljava/util/Map;, instantiatedMethodType=(Lorg/projectnessie/model/Reference;)Ljava/util/Map;, numCaptured=3])
	- writeReplace data (class: java.lang.invoke.SerializedLambda)
	- object (class org.projectnessie.gc.base.DistributedIdentifyContents$$Lambda$1893/0x0000000800e7e840, org.projectnessie.gc.base.DistributedIdentifyContents$$Lambda$1893/0x0000000800e7e840@685e5967)
	- element of array (index: 0)
	- array (class [Ljava.lang.Object;, size 1)
	- field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
	- object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.api.java.JavaPairRDD$, functionalInterfaceMethod=scala/Function1.apply:(Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/api/java/JavaPairRDD$.$anonfun$toScalaFunction$1:(Lorg/apache/spark/api/java/function/Function;Ljava/lang/Object;)Ljava/lang/Object;, instantiatedMethodType=(Ljava/lang/Object;)Ljava/lang/Object;, numCaptured=1])
	- writeReplace data (class: java.lang.invoke.SerializedLambda)
	- object (class org.apache.spark.api.java.JavaPairRDD$$$Lambda$1894/0x0000000800e80040, org.apache.spark.api.java.JavaPairRDD$$$Lambda$1894/0x0000000800e80040@59b43e35)
	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:41)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:47)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:101)
```
and trying to debug.

File: model/src/main/java/org/projectnessie/model/Branch.java [hunk 1]
```diff
@@ -31,7 +32,7 @@
 @JsonSerialize(as = ImmutableBranch.class)
 @JsonDeserialize(as = ImmutableBranch.class)
 @JsonTypeName("BRANCH")
-public interface Branch extends Reference {
+public interface Branch extends Reference, Serializable {
```
Comment (tomekl007) [inline model/src/main/java/org/projectnessie/model/Branch.java:35]: we need to keep the `Serializable` on this class because Kryo for some reason requires that the class uses as a key in the map implements `Serializable`. 
The kryo is used for all `ImmutableBranch` read/writes when running spark. I've verified it by validating that those logs:
[ImmutableBranchSerializer#read](https://github.com/projectnessie/nessie/pull/3599/files#diff-19a139a5d1fa76b7773b67373a4f3cca5002f712d4b13e471742d5d99b7db933R44) 
a
https://github.com/projectnessie/nessie/pull/3599/files#diff-19a139a5d1fa76b7773b67373a4f3cca5002f712d4b13e471742d5d99b7db933R36
are executed for each object

File: gc/gc-base/src/main/java/org/projectnessie/gc/serialization/ReferencesKryoRegistrator.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    // needed for all types of collections. It will use serializers from
+    // `com.esotericsoftware.kryo.serializers.DefaultSerializers`
+    kryo.register(HashMap.class);
+    kryo.register(HashSet.class);
+    kryo.register(ConcurrentHashMap.class);
+    kryo.register(LinkedHashMap.class);
+    kryo.register(Collections.EMPTY_LIST.getClass());
+    kryo.register(Collections.EMPTY_MAP.getClass());
+    kryo.register(Collections.EMPTY_SET.getClass());
+    kryo.register(Collections.singletonList("").getClass());
```
Comment (tomekl007) [inline gc/gc-base/src/main/java/org/projectnessie/gc/serialization/ReferencesKryoRegistrator.java:84]: it is executed only once. Doing `Collections.singletonList("")` is a way to alleviate the problem that the inner class it private (same for two lines below)

File: gc/gc-base/src/main/java/org/projectnessie/gc/serialization/ImmutableGCParamsSerializer.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+public class ImmutableGCParamsSerializer extends Serializer<ImmutableGCParams> {
+  public ImmutableGCParamsSerializer() {
+    this.setImmutable(true);
+  }
+
+  @Override
+  public void write(Kryo kryo, Output output, ImmutableGCParams immutableGCParams) {
+    kryo.writeClassAndObject(output, immutableGCParams.getNessieClientConfigs());
+    kryo.writeClassAndObject(output, immutableGCParams.getCutOffTimestampPerRef());
```
Comment (tomekl007) [inline gc/gc-base/src/main/java/org/projectnessie/gc/serialization/ImmutableGCParamsSerializer.java:35]: `writeClassAndObject` is used when we need to save a type of the object - it is useful for parameterized types like maps.

File: gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java [hunk 4]
```diff
@@ -44,31 +43,32 @@
  * Contains the methods that executes in spark executor for {@link
  * GCImpl#identifyExpiredContents(SparkSession)}.
  */
-public class IdentifyContentsPerExecutor implements Serializable {
+public class IdentifyContentsPerExecutor {
 
-  private final GCParams gcParams;
```
Comment (tomekl007) [inline gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:49]: I've removed the state from this class and changed processing to `static` methods.
This is a favorable way to do processing in spark as it reduces the number of code that needs to be serialized.

File: gc/gc-base/pom.xml [hunk 5]
```diff
@@ -41,6 +41,12 @@
       <version>${project.version}</version>
       <scope>provided</scope>
     </dependency>
+    <!--- We cannot use version > 0.43 beacuse it uses a kryo version 5.x that is incompatible with out spark (requires 4.x) -->
```
Comment (tomekl007) [inline gc/gc-base/pom.xml:44]: it provides needed  `ImmutableMapSerializer` and `UnmodifiableCollectionsSerializer`

File: model/src/main/java/org/projectnessie/model/Branch.java (hunk 1 above)
Comment (ajantha-bhat) [inline model/src/main/java/org/projectnessie/model/Branch.java:35]: Keeping classes of `model` as `Serializable` is not good. Hence, I used JSON string instead of actual class for the references.

I think same problem comes for `Tag` also, currently we don't have test case that creates a tag and drops it and use it for GC. Hence, it looks like problem is not there for `Tag` 

Also keeping `Serializable` with `Immutables` is a problem with spark code generation (which happens with dataframe), code generation will fail because these Immutables classes don't have public consutructor, cannot add public constructor as we are using `@Value.Check` annotation.

File: gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java [hunk 6]
```diff
@@ -31,30 +32,30 @@
...
+   * @param references list of all the references
    * @param bloomFilterSize size of bloom filter to be used
-   * @param droppedRefTimeMap map of dropped time for reference@hash (JSON serialized)
+   * @param droppedRefTimeMap map of dropped time for reference@hash
    * @return map of {@link ContentBloomFilter} per content-id.
    */
-  public Map<String, ContentBloomFilter> getLiveContentsBloomFilters(
-      List<String> references, long bloomFilterSize, Map<String, Instant> droppedRefTimeMap) {
-    IdentifyContentsPerExecutor executor = new IdentifyContentsPerExecutor(gcParams);
+  // here
```
Comment (ajantha-bhat) [inline gc/gc-base/src/main/java/org/projectnessie/gc/base/DistributedIdentifyContents.java:44]: revert this ?

File: gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java [hunk 7]
```diff
@@ -95,17 +81,11 @@ public GCImpl(GCParams gcParams) {
    * @param session spark session for distributed computation
    * @return {@link IdentifiedResult} object having expired contents per content id.
    */
-  public IdentifiedResult identifyExpiredContents(SparkSession session) {
+  public static IdentifiedResult identifyExpiredContents(SparkSession session, GCParams gcParams) {
```
Comment (ajantha-bhat) [inline gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:84]: I would probably keep refactoring in a separate PR.

Also as this flow is WIP, may be better to wait till we merge #3421

**Review by ajantha-bhat — CHANGES_REQUESTED** [review]
Thanks for working on this problem.

But we are still keeping `Serializable` which causes problem for spark code generation with these `Immutables`.  
I too don't have solution at the moment. That is why I have used a JSON serialization instead of Kryo.

File: model/src/main/java/org/projectnessie/model/Branch.java (hunk 1 above)
Comment (tomekl007) [reply model/src/main/java/org/projectnessie/model/Branch.java:35]: yes, using immutables is problematic. One solution would be to create separate data classes that are used only in spark and create a mapper logic between Spark classes and immutables.

File: gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java (hunk 7 above)
Comment (tomekl007) [reply gc/gc-base/src/main/java/org/projectnessie/gc/base/GCImpl.java:84]: Those changes are needed for serialization changes, so it cannot be split.

File: model/src/main/java/org/projectnessie/model/Branch.java (hunk 1 above)
Comment (ajantha-bhat) [reply model/src/main/java/org/projectnessie/model/Branch.java:35]: How efficient it would be compared current JSON serialization? I agree that Kryo is faster but this code is looking complex and adding mapper will bring extra complexity. 

So, I am happy with the current JSON serialization considering the complexity. Let's wait for the others opinion.

Comment (tomekl007) [reply model/src/main/java/org/projectnessie/model/Branch.java:35]: The kryo should be orders of magnitude faster compared to json (see https://blog.softwaremill.com/the-best-serialization-strategy-for-event-sourcing-9321c299632b) and also it requires a lot less space for storing each object in memory, so I think we should definitely use kryo instead of JSON.

Comment (tomekl007) [reply model/src/main/java/org/projectnessie/model/Branch.java:35]: In Spark, it is a good pattern to model the domain using case classes (when writing in scala) and carry only the parts of data that are indeed needed for processing. Since we are using Java we cannot use case classes but we can model our problem and keep it contained to the spark module. This is the same pattern as for rest APIs and DTOs. The entity (persistence) model should not leak into a REST api that is why the DTOs are used. For big data processing, we should also have a dedicated model classes to not tight couple spark processing to our model.

**Comment by ajantha-bhat** [comment]
@tomekl007: Also please wait till https://github.com/projectnessie/nessie/pull/3421 get merged, I am removing `ContentValues` and `IdentifiedResult` there. So, we can avoid some rework.

File: gc/gc-base/pom.xml [hunk 8]
```diff
@@ -41,6 +41,12 @@
...
+    <dependency>
+      <groupId>de.javakaffee</groupId>
+      <artifactId>kryo-serializers</artifactId>
+      <version>0.43</version>
```
Comment (nastra) [inline gc/gc-base/pom.xml:48]: can we move this out into a property within the root `pom.xml` and also define the dependency in the `dependencyManagement` section?

File: gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java (hunk 4 above)
Comment (nastra) [reply gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:49]: not sure I fully understand why moving things to `static` reduces the amount of code that needs to be serialized. Can you elaborate here a bit please?

File: gc/gc-base/src/test/java/org/projectnessie/gc/serialization/ReferencesSerializerTest.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+    // given
+    Map<String, String> properties = new HashMap<>();
+    properties.put("key", "value");
+    properties.put("key2", "value2");
+
+    Kryo kryo = new Kryo();
+    new ReferencesKryoRegistrator().registerClasses(kryo);
+    ImmutableBranch immutableBranch =
+        ImmutableBranch.builder()
+            .hash("0a62a2d020df785dfb3d34e9fb64d965e6e65220")
```
Comment (nastra) [inline gc/gc-base/src/test/java/org/projectnessie/gc/serialization/ReferencesSerializerTest.java:52]: I believe this needs a few more tests where for example everything that's `Nullable` in `Branch` is actually omitted when creating a `Branch` instance.

Given that, I think it would be good if there would be a separate test class for each custom serializer and within that test class we can then make sure all edge cases are covered

**Review by nastra — CHANGES_REQUESTED** [review]

File: model/src/main/java/org/projectnessie/model/Branch.java (hunk 1 above)
Comment (tomekl007) [reply model/src/main/java/org/projectnessie/model/Branch.java:35]: The `Spark` end-to-end test: `TestJerseyRestGcInMemory`

File: gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java (hunk 4 above)
Comment (tomekl007) [reply gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java:49]: When we have a spark processing and there is an instance variable on the spark class (state on the class) the whole class needs to be serialized and sent by the spark to executors (https://stackoverflow.com/questions/40818001/understanding-spark-serialization) - for that reason, it is a common pattern in spark to have stateless processing to avoid the need of serialization.

File: gc/gc-base/pom.xml (hunk 8 above)
Comment (tomekl007) [reply gc/gc-base/pom.xml:48]: do you want me to move all dependencies from this module to the new `dependencyManagement` section or only this one?

File: gc/gc-base/src/test/java/org/projectnessie/gc/serialization/ReferencesSerializerTest.java (hunk 9 above)
Comment (tomekl007) [reply gc/gc-base/src/test/java/org/projectnessie/gc/serialization/ReferencesSerializerTest.java:52]: agree, I wanted to wait until we will settle out whether we want to use the kryo or not. If we decide that this is a good direction, I will add more unit tests.

File: gc/gc-base/pom.xml [hunk 10]
```diff
@@ -29,6 +29,11 @@
 
   <name>Nessie - GC - Base Implementation</name>
 
+  <properties>
+    <!--- We cannot use version > 0.43 beacuse it uses a kryo version 5.x that is incompatible with out spark (requires 4.x) -->
+    <kryo.serializers.version>0.43</kryo.serializers.version>
```
Comment (nastra) [inline gc/gc-base/pom.xml:34]: should rather go into https://github.com/projectnessie/nessie/blob/main/pom.xml and the dependency + version should be defined in https://github.com/projectnessie/nessie/blob/main/pom.xml#L223. Then you can omit the `<version>${kryo.serializers.version}</version>` field when you define the dependency in this module here

**Comment by tomekl007** [comment]
> I think it would be good if there would be a separate test class for each custom serializer and within that test class we can then make sure all edge cases are covered (like omitting nullable fields, ....)

Agree, see: https://github.com/projectnessie/nessie/pull/3599#discussion_r827740293

**Comment by snazy** [comment]
Is this one still needed?
